homeassistant/components/minecraft_server/__init__.py

```diff
@@ -62,6 +71,105 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
+    _LOGGER.debug("Migrating from version %s", config_entry.version)
+
+    # 1 --> 2: Use config entry ID as base for unique IDs.
+    if config_entry.version == 1:
+        assert config_entry.unique_id
+        assert config_entry.entry_id
+        old_unique_id = config_entry.unique_id
+        new_unique_id = config_entry.entry_id
+
+        # Migrate config entry.
+        _LOGGER.debug(
+            "Migrating config entry unique ID from %s to %s",
+            old_unique_id,
+            new_unique_id,
+        )
+        config_entry.unique_id = new_unique_id
+        config_entry.version = 2
+        hass.config_entries.async_update_entry(config_entry)
+
+        # Migrate device.
+        await _async_migrate_device_identifiers(hass, config_entry, old_unique_id)
+
+        # Migrate entities.
+        await er.async_migrate_entries(
+            hass, config_entry.entry_id, _migrate_entity_unique_id
+        )
+
+    _LOGGER.info("Migration to version %s successful", config_entry.version)
+
+    return True
+
+
+async def _async_migrate_device_identifiers(
+    hass: HomeAssistant, config_entry: ConfigEntry, old_unique_id: str | None
+) -> None:
+    """Migrate the device identifiers to the new format."""
+    device_registry = dr.async_get(hass)
+    for device_entry in dr.async_entries_for_config_entry(
+        device_registry, config_entry.entry_id
+    ):
+        assert device_entry
+        for identifier in device_entry.identifiers:
+            if identifier[1] == old_unique_id:
+                # Device found in registry. Update identifiers.
+                new_identifiers = {
+                    (
+                        DOMAIN,
+                        config_entry.entry_id,
+                    )
+                }
+                _LOGGER.debug(
+                    "Migrating device identifiers from %s to %s",
+                    device_entry.identifiers,
+                    new_identifiers,
+                )
+                device_registry.async_update_device(
+                    device_entry.id, new_identifiers=new_identifiers
+                )
+                # Leave for loop, as each server has only one device.
+                break
+
+
+@callback
+def _migrate_entity_unique_id(entity_entry: er.RegistryEntry) -> dict[str, Any]:
+    """Migrate the unique ID of an entity to the new format."""
+    assert entity_entry
+
+    # Different variants of unique IDs are available in version 1:
+    # 1) SRV record: '<host>-srv-<entity_type>'
+    # 2) Host & port: '<host>-<port>-<entity_type>'
+    # 3) IP address & port: '<mac_address>-<port>-<entity_type>'
+    unique_id_pieces = entity_entry.unique_id.split("-")
+    entity_type = unique_id_pieces[2]
+
+    # Handle bug in version 1: Entity type names were used instead of
+    # keys (e.g. "Protocol Version" instead of "protocol_version").
+    new_entity_type = entity_type.lower()
+    new_entity_type = new_entity_type.replace(" ", "_")
```

Comment:
I was about to say this on your other PR. Good change!

Reply:
Thanks. I was surprised that spaces in unique IDs don't break stuff.

Reply:
Me too but we tend to want to use `-` and not `_`

Reply:
I can change the delimiter of the entity types from underscore to dash. But this would also affect the translation keys, as I‘m using the same constants from const.py. Should I change it for both (entity types and translation keys)?

I was not aware of it, because it‘s not written anywhere (at least I couldn‘t find it) and I some integrations I checked are all using underscore.

Reply:
Translation keys have to use underscore afaik

Reply:
I think that's the best way forward yes :)

Reply:
Done.

Reply:
I did not check the code really but I hope we have a test here that since we are changing the constants that this really work as it's not only changing the prefix anymore but the the suffix (the constant).
If we feel safer then leave it with underscore and just use the same constant for translation key if preferred. It's a matter of personal taste I guess.

Reply:
After each change I always test two things:
- Adding an instance of the integration in my branch.
- Adding an instance of the integration in the dev branch and then switch to my branch to trigger the migration.

If I didn't oversee something, then the change of the suffix looks good. 

My preference was to use the constant for both, translation keys and entity unique ID suffixes. But I'll do whatever is the preferred way from HA core point of view.

Reply:
Let's use `KEY_PLAYERS_MAX = "players-max"` as example
Then the unique id should look like `3454326455324423-players_max`

So your constant should be `KEY_PLAYERS_MAX = "players_max"` which will also suit translation key.
The `-` is only to separate the entitiy_id and the constant in this case (which is what I meant above. Replacing spaces is normally done with `_`.

Reply:
Oooh, I'm sorry, I didn't understand it like this.

Reply:
This is exactly how I implemented it in the first place, classic misunderstanding 😄 

I‘ll revert my latest commit then…

Reply:
Done.